### PR TITLE
Support tctl through a proxy

### LINKED ISSUE #3431: Let `tctl` control Teleport Auth via the Proxy Server.
**What happened**:
While testing https://github.com/gravitational/teleport/issues/3018 I found that it only works for auth servers that are publicly accessible. I often run the auth server on a private subnet, as it doesn't have easy internet ingress. This proposal is the same as our IoT feature and would let admins access the auth server via the proxy server. 

**What you expected to happen**:

```
remote$ tctl auth sign \
                --auth-server=teleport-proxy.example.com:3080 \
                --identity=identity.pem \
                --user=alice \
                --out=alice.pem 
```

### PR DESCRIPTION
**Pull request by awly**
Use the reverse tunnel endpoint, similar to IoT nodes, to connect to the
auth server. Also add an `--insecure` flag, similar to tsh, for testing
with self-signed certs on the proxy.

Fixes https://github.com/gravitational/teleport/issues/3431

### CODE REVIEW

**Comment by benarent** [comment]
To double check, the customer UX will be similar to below. Copied from https://github.com/gravitational/teleport/issues/3018 

To use it remotely, a user would first need to export identity locally on the auth server using the usual method:
```
auth$ tctl auth sign --user=admin --out=identity.pem
```
After that, the exported identity file can be used to invoke tctl from a remote node:

```
remote$ tctl status \
                --auth-server=teleport.example.com:3080 \
                --identity=identity.pem
```

**Comment by webvictim** [comment]
We might want to be explicitly clear that this doesn't work with the regular tsh identity files that are stored under `~/.tsh/keys/<cluster>/<username>` after you log in. That was the first thing I tried and was surprised to get errors:

```
$ ~/go/src/github.com/gravitational/teleport/build/tctl -i ~/.tsh/keys/teleport.example.com/exampleuser --auth-server=teleport.example.com:3080 users
INFO[0000] Reading identity file: /home/gus/.tsh/keys/teleport.example.com/exampleuser 
INFO[0000] Certificate not found in /home/gus/.tsh/keys/teleport.example.com/exampleuser . Looking in /home/gus/.tsh/keys/teleport.example.im/exampleuser -cert.pub.

ERROR REPORT:
Original Error: *errors.errorString tls: failed to find any PEM data in certificate input
Stack Trace:
        /home/gus/go/src/github.com/gravitational/teleport/lib/client/interfaces.go:110 github.com/gravitational/teleport/lib/client.(*Key).ClientTLSConfig
        /home/gus/go/src/github.com/gravitational/teleport/tool/tctl/common/tctl.go:338 github.com/gravitational/teleport/tool/tctl/common.applyConfig
        /home/gus/go/src/github.com/gravitational/teleport/tool/tctl/common/tctl.go:132 github.com/gravitational/teleport/tool/tctl/common.Run
        /home/gus/go/src/github.com/gravitational/teleport/tool/tctl/main.go:35 main.main
        /usr/local/go/src/runtime/proc.go:212 runtime.main
        /usr/local/go/src/runtime/asm_amd64.s:1374 runtime.goexit
User Message: failed to parse TLS cert and key
```

**Comment by awly** [comment]
@webvictim yeah, it needs both TLS and SSH certs i think (for auth server and proxy tunnel respectively).
In theory, it could read the profile directory and get those from individual files.

For now, let's just document the limitation. If users demand to use their `tsh` creds with `tctl`, we'll add that support.
@benarent do you want to write the docs for this or should I draft something in this PR?

**Review by webvictim — APPROVED** [review]
Going to approve this because it worked for me.

File: lib/client/interfaces.go [hunk 1]
```diff
@@ -288,3 +308,34 @@ func (k *Key) CheckCert() error {
...
+		if err != nil {
+			return nil, trace.BadParameter("failed parsing CA cert: %v; raw CA cert line: %q", err, caCert)
+		}
+		trustedKeys = append(trustedKeys, publicKey)
+	}
+	// No CAs are provided, return a nil callback which will prompt the user
+	// for trust.
+	if len(trustedKeys) == 0 {
+		return nil, nil
+	}
```
Comment (fspmarshall) [inline lib/client/interfaces.go:327]: Unexpected behavior for a public API.  Either document this behavior very explicitly in function doc comments or return a `NotFoundError` and then handle it explicitly at the points where a `nil` value for `HostKeyCallback` is OK.

**Review by fspmarshall — APPROVED** [review]

File: lib/utils/testlog/log.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+	"testing"
+
+	"github.com/sirupsen/logrus"
+)
+
+// FailureOnly returns a logger that only prints the logs to STDERR when the
+// test fails.
+func FailureOnly(t *testing.T) *logrus.Entry {
+	// Collect all the output in buf.
+	buf := &bytes.Buffer{}
```
Comment (klizhentas) [inline lib/utils/testlog/log.go:17]: Use `utils.SyncBuffer` here to prevent data races on concurrent logs?

File: lib/utils/testlog/log.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	"fmt"
+	"os"
+	"testing"
+
+	"github.com/sirupsen/logrus"
+)
+
+// FailureOnly returns a logger that only prints the logs to STDERR when the
+// test fails.
+func FailureOnly(t *testing.T) *logrus.Entry {
```
Comment (klizhentas) [inline lib/utils/testlog/log.go:15]: we have `utils.InitLoggerForTests(testing.Verbose())` Does it make sense to use `FailureOnly` logger in that function too by default?

File: tool/tctl/common/tctl.go [hunk 4]
```diff
@@ -154,73 +173,138 @@ func connectToAuthService(cfg *service.Config) (client auth.ClientI, err error)
...
 	if err != nil {
-		utils.Consolef(os.Stderr, teleport.ComponentClient,
-			"Cannot connect to the auth server: %v.\nIs the auth server running on %v?",
-			err, cfg.AuthServers[0].Addr)
-		os.Exit(1)
+		err = trace.Wrap(err, "failed direct dial to auth server: %v", err)
+		if clientConfig.sshConfig == nil {
+			// No identity file was provided, don't try dialing via a reverse
+			// tunnel on the proxy.
+			return nil, err
```
Comment (klizhentas) [inline tool/tctl/common/tctl.go:191]: `trace.Wrap`

File: tool/tsh/common/identity.go [hunk 5]
```diff
@@ -60,66 +54,52 @@ func LoadIdentity(idFn string) (*client.Key, ssh.HostKeyCallback, error) {
...
-				hostKey = clusterCert.SignatureKey
-			}
-			for _, trustedKey := range trustedKeys {
-				if sshutils.KeysEqual(trustedKey, hostKey) {
-					return nil
-				}
+		for _, caCert := range ident.CACerts.SSH {
+			_, _, _, _, _, err := ssh.ParseKnownHosts(caCert)
+			if err != nil {
+				return nil, trace.BadParameter("CA cert parsing error: %v. cert line :%v", err.Error(), string(caCert))
```
Comment (klizhentas) [inline tool/tsh/common/identity.go:101]: can we tell user what to do with this error?

File: lib/utils/testlog/log.go (hunk 3 above)
Comment (awly) [reply lib/utils/testlog/log.go:15]: `FailureOnly` depends on per-test state (`testing.T`)
`utils.InitLoggerForTests` is inherently global and doesn't have access to per-test info.
I can't think of a way to implement something similar to `utils.InitLoggerForTests` using `FailureOnly`

**Pull request merged**